NFV OVS DPDK document changes (#111)
openstack-k8s-operators/dev-docs: nfv_ovs_dpdk.md

# NFV OVS DPDK

This document describes how to deploy EDPM nodes with OVS DPDK.

## Procedure Overview

1. Configure kernel args and tuned parameters
2. Configure OVS DPDK parameters
3. Configure the networks with DPDK interfaces of the EDPM nodes
2. Configure Nova and Neutron parameters.

In order to complete the above procedure, the `services` list of the
`OpenStackDataPlaneNodeSet` CR needs to be edited.

## OpenStackDataPlaneNodeSet services list

EDPM nodes can be configured by creating an
`OpenStackDataPlaneNodeSet` CR which the
[dataplane-operator](https://openstack-k8s-operators.github.io/dataplane-operator)

> will reconcile when an `OpenStackDataPlaneDeployment` CR is created.

Reviewer: I would say `will deploy' instead of `will reconcile`, since technically the `OpenStackDataPlaneNodeSet` is reconciled even without an `OpenStackDataPlaneDeployment` (it just won't process certain things yet).  Or maybe we can say something like `will reconcile to create OpenStackDataPlaneService resources`, because I _think_ the services are created when the `OpenStackDataPlaneDeployment` is created.

Author: ack

These types of CRs have a `services` list like the following:

```yaml
apiVersion: dataplane.openstack.org/v1beta1
kind: OpenStackDataPlaneNodeSet
spec:
  ...
  services:
    - bootstrap
    - download-cache
    - reboot-os
    - configure-ovs-dpdk
    - configure-network
    - validate-network
    - install-os
    - configure-os
    - ssh-known-hosts
    - run-os
    - install-certs
    - ovn
    - neutron-ovn
    - neutron-metadata
    - libvirt
    - nova-custom-ovsdpdk
```
Only the services which are on the list will be configured.

## Configure the EDPM ansible variables of the EDPM nodes

Following are the list of EDPM ansible variables which need to be
provided for deploying with OVS DPDK support.

edpm_ovs_dpdk_pmd_core_list: List of Logical CPUs to be allocated for Poll Mode Driver 

edpm_ovs_dpdk_socket_memory: Socket memory list per NUMA node

edpm_ovs_dpdk_memory_channels: Number of memory channel per socket

edpm_ovn_bridge_mappings: List of bridge and dpdk ports mappings

edpm_kernel_args: Configures kernel boot parameters

edpm_tuned_profile: tuned profile configuration

edpm_tuned_isolated_cores: List of isolated cores

> edpm_nova_libvirt_qemu_group: Configures qemu group for nova libvirt

Reviewer: Maybe put ticks around each variable here i.e. `edpm_ovs_dpdk_pmd_core_list`

Example:
dpm_ovs_dpdk_pmd_core_list: "1,13,2,14,3,15"
edpm_ovs_dpdk_socket_memory: "4096"
edpm_ovs_dpdk_memory_channels: "4"
edpm_ovn_bridge_mappings: ['dpdk2:br-link2','dpdk1:br-link1']
edpm_kernel_args: "default_hugepagesz=1GB hugepagesz=1G hugepages=64 iommu=pt intel_iommu=on tsx=off isolcpus=2-11,14-23"
edpm_nova_libvirt_qemu_group: "hugetlbfs"
edpm_tuned_profile: "cpu-partitioning"

> edpm_tuned_isolated_cores: "2-11,14-23"

Reviewer: Maybe put this whole block in triple-ticks

Author: done

> edpm_network_config_template:

Reviewer: `edpm_network_config_template` (i.e. add ticks)

Author: > `edpm_network_config_template` (i.e. add ticks)

done

Do not yet create the CR in OpenShift as the edits described in the
next sections are required.

The example
[dataplane_v1beta1_openstackdataplanenodeset_ovs_dpdk.yaml]https://github.com/openstack-k8s-operators/dataplane-operator/tree/main/examples/ovs_dpdk)
has OVS DPDK interfaces network configuration and required edpm ansible parameters.

Modify your `OpenStackDataPlaneNodeSet` CR to set
[edpm-ansible](https://github.com/openstack-k8s-operators/edpm-ansible)
variables so that the
[edpm_network_config role](https://github.com/openstack-k8s-operators/edpm-ansible/blob/main/roles/edpm_network_config/)
will configure networks with ovs dpdk interfaces.

## Configure Nova

Create a ConfigMap with content to be added to /etc/nova/nova.conf.d/,
```yaml
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: ovs-dpdk-cpu-pinning-nova
data:
  04-cpu-pinning-nova.conf: |
    [DEFAULT]
    reserved_host_memory_mb = 4096
    [compute]
    cpu_shared_set = 0-3,24-27
    cpu_dedicated_set = 8-23,32-47
    [neutron]
    physnets = dpdk1, dpdk2
    [neutron_physnet_dpdk1]
    numa_nodes = 0
    [neutron_physnet_dpdk2]
    numa_nodes = 0
    [neutron_tunnel]
    numa_nodes = 0
```

> The filename, e.g. 04-cpu-pinning-nova.conf, must match "*nova*.conf" and
> the files are evaluated by Nova alphabetically (e.g. 01-foo-nova.conf
> is processed before 02-bar-nova.conf).

Reviewer: ```suggestion
The filename, e.g. `04-cpu-pinning-nova.conf`, must match `*nova*.conf` and
the files are evaluated by Nova alphabetically (e.g. `01-foo-nova.conf`
is processed before `02-bar-nova.conf`).
```

Author: > 

done

  configMaps:
    - ovs-dpdk-cpu-pinning-nova
  secrets:
    - nova-cell1-compute-config
  playbook: osp.edpm.nova
```
The custom service is named `nova-custom-ovsdpdk`. It cannot be named
`nova` because `nova` is an immutable default service and will
overwrite any custom service with the same name during reconciliation.

After the `ConfigMap` and `OpenStackDataPlaneService` services above
have been created (e.g. `oc create -f nova-custom-ovsdpdk.yaml`), update the
`OpenStackDataPlaneNodeSet`
[EDPM services list](https://openstack-k8s-operators.github.io/dataplane-operator/composable_services)
to replace the `nova` service with `nova-custom-ovsdpdk`.

```yaml
apiVersion: dataplane.openstack.org/v1beta1
kind: OpenStackDataPlaneNodeSet
spec:
  ...
  roles:
    edpm-compute:
      ...
      services:
        - bootstrap
        - download-cache
        - reboot-os
        - configure-ovs-dpdk
        - configure-network
        - validate-network
        - install-os
        - configure-os
        - ssh-known-hosts
        - run-os
        - install-certs
        - ovn
        - neutron-ovn
        - neutron-metadata
        - libvirt
        - nova-custom-ovsdpdk
```

> The `bootstrap`, 'configure-ovs-dpdk' and 'configure-network' services

Reviewer: ```suggestion
The `bootstrap`, `configure-ovs-dpdk` and `configure-network` services
```

Author: > 

done

overrides from the ConfigMaps onto the Nova hosts.

### Create the OpenStackDataPlaneNodeSet

Create the CR from your directory based on the example
[dataplane_v1beta1_openstackdataplanenodeset_ovs_dpdk](https://github.com/openstack-k8s-operators/dataplane-operator/tree/main/examples/ovs_dpdk)
with the changes described in the previous section.
```
oc kustomize --load-restrictor LoadRestrictionsNone dataplane-operator/examples/ovs_dpdk > dataplane_cr.yaml
```

### Create a OpenStackDataPlaneDeployment

Creating an `OpenStackDataPlaneDeployment` will trigger Ansible jobs
to configure an EDPM. Which Ansible roles are run depends on the
`services` list.

Each `OpenStackDataPlaneDeployment` can have its own
`servicesOverride` list which will redefine the list
of services of an `OpenStackDataPlaneNodeSet` for a
deployment.

The example
[dataplane_v1beta1_openstackdataplanedeployment_ovs_dpdk](https://github.com/openstack-k8s-operators/dataplane-operator/tree/main/examples/ovs_dpdk)

Create the CR based on the example
```
oc kustomize --load-restrictor LoadRestrictionsNone dataplane-operator/examples/ovs_dpdk > dataplane_cr.yaml
```

> custom `OpenStackDataPlaneService` called `nova-custom-ovsdpdk`

Reviewer: ```suggestion
Custom `OpenStackDataPlaneService` called `nova-custom-ovsdpdk`
```

Author: > 

done

> has been created as described in
> the [documentation to configure OpenStack to use ](ovs_dpdk.md).

Reviewer: ```suggestion
has been created as described in
the [documentation to configure OpenStack to use OVS-DPDK](ovs_dpdk.md).
```

Author: > 

done

The `nova-custom-dpdk` can be seen in the
[example](https://github.com/openstack-k8s-operators/dataplane-operator/tree/main/examples/ovs_dpdk)
and takes the place of the default `nova`

> OpenStackDataPlaneService. This custom service uses a ConfigMap called

Reviewer: ```suggestion
`OpenStackDataPlaneService`. This custom service uses a `ConfigMap` called
```

Author: > 

done

# NFV OVS DPDK

This document describes how to deploy EDPM nodes with OVS DPDK.

## Procedure Overview

1. Configure kernel args and tuned parameters
2. Configure OVS DPDK parameters
3. Configure the networks with DPDK interfaces of the EDPM nodes
2. Configure Nova and Neutron parameters.

In order to complete the above procedure, the `services` list of the
`OpenStackDataPlaneNodeSet` CR needs to be edited.

## OpenStackDataPlaneNodeSet services list

EDPM nodes can be configured by creating an
`OpenStackDataPlaneNodeSet` CR which the
[dataplane-operator](https://openstack-k8s-operators.github.io/dataplane-operator)
will reconcile to create OpenStackDataPlaneService resources
when an `OpenStackDataPlaneDeployment` CR is created.

> These types of CRs have a `services` list like the following:

Author: done

```yaml
apiVersion: dataplane.openstack.org/v1beta1
kind: OpenStackDataPlaneNodeSet
spec:
  ...
  services:
    - bootstrap
    - download-cache
    - reboot-os
    - configure-ovs-dpdk
    - configure-network
    - validate-network
    - install-os
    - configure-os
    - ssh-known-hosts
    - run-os
    - install-certs
    - ovn
    - neutron-ovn
    - neutron-metadata
    - libvirt
    - nova-custom-ovsdpdk
```
Only the services which are on the list will be configured.

## Configure the EDPM ansible variables of the EDPM nodes

Following are the list of EDPM ansible variables which need to be

> provided for deploying with OVS DPDK support.

Author: > Link to roles involved in the setup would provide more up to date information about variables involved. Especially since the roles may change in the future and they may also contain other information about nature of "services" deployed.

done
